File: src/slx.c
#include "slope.h"

/* -------------------------- slx -------------------------- */
static t_class *slx_class;

static void slope_k(t_slope *x) {
	if (x->log)
	{    slope_minmax(x);
	     x->k = x->run / log(x->max / x->min);  }
	else x->k = x->run /    (x->max - x->min);
}

static void slx_float(t_slope *x ,t_float f) {
	t_float res = (x->log)
		? log(f / x->min) * x->k
		:    (f - x->min) * x->k;
	outlet_float(x->obj.ob_outlet ,res);
}

static void *slx_new(t_symbol *s ,int argc ,t_atom *argv) {
	return slope_new(slx_class ,s ,argc ,argv);
}

void slx_setup(void) {
	slx_class = slope_setup(gensym("slx") ,(t_newmethod)slx_new);
	class_addfloat(slx_class ,slx_float);
	class_sethelpsymbol(slx_class ,gensym("slope"));
}

File: src/slx.zig
const pd = @import("pd");
const Slope = @import("slope.zig").Slope(@This());
pub const name = "slx";

pub inline fn getK(min: f64, max: f64, run: f64) f64 {
	return run / (max - min);
}

pub fn floatC(p: *const pd.Pd, f: pd.Float) callconv(.c) void {
	const self = Slope.parentConstPtr(p);
	self.out.float(@floatCast((f - self.min) * self.k));
}

export fn slx_setup() void {
	_ = pd.wrap(void, Slope.setup(), @src().fn_name);
}
